File: Backend/StudentManagement.Domain/Models/Course.cs
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace StudentManagement.Domain.Models
{
    [Table("courses")]
    [Index(nameof(CourseName), IsUnique = true)]
    public class Course
    {
        [Column("id", TypeName ="varchar(10)")]
        [Key]
        public string CourseId { get; set; } = null!;

        [Column("name", TypeName ="nvarchar(50)")]
        public string CourseName { get; set; } = null!;

        [Column("name_eng", TypeName = "nvarchar(50)")]
        public string CourseNameEng { get; set; } = null!;

        [Column("credits")]
        [Range(2, int.MaxValue)]
        public int Credits { get; set; }

        [Column("faculty_id")]
        public Guid FacultyId { get; set; }
        public Faculty Faculty { get; set; } = null!;

        [Column("description", TypeName = "nvarchar(max)")]
        public string Description { get; set; } = string.Empty;

        [Column("description_eng", TypeName = "nvarchar(max)")]
        public string DescriptionEng { get; set; } = string.Empty;
        public Course? RequiredCourse { get; set; }

        [Column("required_course_id")]
        public string? RequiredCourseId { get; set; }

        [Column("deleted_at")]
        public DateTime? DeletedAt { get; set; }
        [Column("created_at")]
        public DateTime CreatedAt { get; set; }

        public ICollection<Class> Classes { get; set; } = null!;
    }
}

File: frontend/src/types/course.ts
import { Faculty } from "./student";

export interface Course {
  id: string;
  courseId: string;
  courseName: {
    vi: string;
    en: string;
  };
  credits: number;
  facultyId: Faculty["id"];
  facultyName?: {
    vi: string;
    en: string;
  };
  description: {
    vi: string;
    en: string;
  };
  requiredCourseId?: string;
  requiredCourseName?: {
    vi: string;
    en: string;
  };
  deletedAt?: string;
  createdAt?: string;
}

export interface CourseGrade {
  code: string;
  name: {
    vi: string;
    en: string;
  };
  credits: number;
  score: number;
  grade: string;
}

export type CourseFormData = {
  code: string;
  nameVi: string;
  nameEn: string;
  credits: number | string;
  faculty: string;
  descriptionVi: string;
  descriptionEn: string;
  prerequisites: string;
};

export interface TranscriptData {
  courses: CourseGrade[];
  gpa: number;
  totalCredits: number;
  completedCredits: number;
}
